src/plonk/lookup/prover.rs

```diff
@@ -327,35 +330,44 @@ impl<C: CurveAffine> Permuted<C> {
         // Compute the evaluations of the lookup product polynomial
         // over our domain, starting with z[0] = 1
         let z = iter::once(C::Scalar::one())
-            .chain(lookup_product.into_iter().skip(1))
+            .chain(lookup_product)
             .scan(C::Scalar::one(), |state, cur| {
                 *state *= &cur;
                 Some(*state)
             })
+            // Take all rows including the "last" row which should
+            // be a boolean (and ideally 1, else soundness is broken)
+            .take(params.n as usize - blinding_factors)
+            // Chain random blinding factors.
+            .chain((0..blinding_factors).map(|_| C::Scalar::rand()))
             .collect::<Vec<_>>();
+        assert_eq!(z.len(), params.n as usize);
         let z = pk.vk.domain.lagrange_from_vec(z);
 
         #[cfg(feature = "sanity-checks")]
         // This test works only with intermediate representations in this method.
         // It can be used for debugging purposes.
         {
             // While in Lagrange basis, check that product is correctly constructed
-            let n = params.n as usize;
+            let n = (params.n as usize) - (blinding_factors + 1);
 
-            // z'(X) (a'(X) + \beta) (s'(X) + \gamma)
-            // - z'(\omega^{-1} X) (\theta^{m-1} a_0(X) + ... + a_{m-1}(X) + \beta) (\theta^{m-1} s_0(X) + ... + s_{m-1}(X) + \gamma)
+            // l_0(X) * (1 - z(X)) = 0
+            assert_eq!(z[0], C::Scalar::one());
+
+            // z(\omega X) (a'(X) + \beta) (s'(X) + \gamma)
+            // - z(X) (\theta^{m-1} a_0(X) + ... + a_{m-1}(X) + \beta) (\theta^{m-1} s_0(X) + ... + s_{m-1}(X) + \gamma)
             for i in 0..n {
-                let prev_idx = (n + i - 1) % n;
+                let next_idx = (i + 1) % n;
 
-                let mut left = z[i];
+                let mut left = z[next_idx];
```

Comment:
```suggestion
                let mut left = z[i+1];
```

Reply:
This uncovered a bug in the sanity check actually, thanks.